_posts/2024-01-25-websockets-in-http4s.md

```diff
@@ -0,0 +1,1561 @@
+
+In the `commandParser` implementation we create `leftSide` which is of type `Parser[String]`, since all commands should start with a `/`, we represent this with `char('/').string` and combine it with `alpha.rep.string` on the right using `~`.
+
+`alpha` on its own is of type `Parser[Char]` and only parses characters containing the alphabet in either lower or uppercase, calling rep transforms the parser to `Parser[NonEmptyList[Char]]`, meaning that we are now parsing one or more characters and finally calling `string` converts means that we can now parse a `String`.
+
+The statement `val leftSide = (char('/').string ~ alpha.rep.string).string` means that `leftSide` parses a string that starts with a `/` followed by characters that contain a-z or A-Z.
+
+The `rightSide` parser is defined by `sp ~ alpha.rep.string` parses strings that start with a space followed by a `string` (a-z or A-Z) and returns a `Parser[(Unit, String)]` since `sp` is of `Parser[Unit]`.
+
+The last section of our parser defined by:
+
+```scala
+((leftSide ~ rightSide.?) <* wsp.rep.?).map((left, right) =>
+  TextCommand(left, right.map((_, s) => s))
+)
+```
+Can be broken down as follows:
+`(leftSide ~ rightSide.?)` means we combine the two parses using the `~` meaning that we keep results from both sides. However the `?` on `rightSide.?` means that rightSide may present or not. In other words, the same parser can also handle commands with only the `leftSide` for example a help command issued as `/help`, this doesn't require an extra parameter.
+
+`<* wsp.rep.?` means if there's white space that follows our command, we ignore it. Finally, we map the left and right values to `TextCommand`, on the right side we ignore the `Unit` value and only map the `String`.
+
+The next private function, `processTextCommand()` takes our command string and passes it to `commandParser.parseAll()` returning an `Either[Parser.Error, TextCommand]`, where `Parser.Error` is returned in case of a parsing failure.
+
+Next, we'll process text values for registered and unregistered users starting with the latter:
+
+```scala
+object InputMessage {
+  ...
+  private def processText4UnReg[F[_]: Monad](
+      text: String,
+      cmd: Command[F],
+      userRef: Ref[F, Option[User]],
+      room: Room
+  ): F[Seq[OutputMessage]] =
+    if (text.charAt(0) == '/') {
+      parseToTextCommand(text).fold(
+        _ =>
+          Seq(
+            ParsingError(
+              None,
+              "Characters after '/' must be between A-Z or a-z"
+            )
+          ).pure[F],
+        v =>
+          v match
+            case TextCommand("/name", Some(n)) =>
+              cmd.isUsernameInUse(n).flatMap { b =>
+                if (b) {
+                  Seq(ParsingError(None, "User name already in use")).pure[F]
+                } else {
+                  cmd.register(n).flatMap {
+                    case SuccessfulRegistration(u, _) =>
+                      userRef
```

Comment:
this is a candidate for a for-comprehension

Reply:
Now uses for comprehension